src/dbnode/storage/database.go

```diff
@@ -453,14 +453,38 @@ func (d *db) Options() Options {
 
 func (d *db) AssignShardSet(shardSet sharding.ShardSet) {
 	d.Lock()
-	defer d.Unlock()
-
 	receivedNewShards := d.hasReceivedNewShardsWithLock(shardSet)
-
-	d.shardSet = shardSet
 	if receivedNewShards {
 		d.lastReceivedNewShards = d.nowFn()
 	}
+	d.Unlock()
+
+	if err := d.mediator.EnqueueMutuallyExclusiveFn(func() {
+		d.assignShardSet(shardSet)
+	}); err != nil {
+		if errors.Is(err, errMediatorNotOpen) {
+			// initial assignment.
+			d.assignShardSet(shardSet)
+		} else {
```

Comment:
Can we special case this rather than return an error and proceed anyway?

i.e.
```go
if !d.mediator.Open() {
  // Initial assignment
  d.assignShardSet(shardSet)
  return
}

if err := d.mediator.EnqueueMutuallyExclusiveFn(func() {
  d.assignShardSet(shardSet)
}); err != nil {
  // log invariant error
}
```